MinecraftClient/McTcpClient.cs

```diff
@@ -1018,5 +1081,298 @@ public void OnPluginChannelMessage(string channel, byte[] data)
+                L.Z += Dz;
+                entitiesToTrack[EntityID].Location = L;
+                if(entitiesToAttack.ContainsKey(EntityID))
+                    entitiesToAttack[EntityID].Location = L;
+                Double distance = calculateDistance(L, GetCurrentLocation());
+                
+                if (distance < attackRange)
+                {
+                    if (!entitiesToAttack.ContainsKey(EntityID))
+                    {
+                        entitiesToAttack.Add(EntityID, entity);
+                    }
+                }
+                else
+                {
+                    entitiesToAttack.Remove(EntityID);
+                }
+            }
+            if (fishingRod.ContainsKey(EntityID))
+            {
+                Location L = fishingRod[EntityID].Location;
+                L.X += Dx;
+                L.Y += Dy;
+                L.Z += Dz;
+                fishingRod[EntityID].Location = L;
+                // check if fishing hook is stationary
+                if (Dx == 0 && Dz == 0)
+                {
+                    if (Dy < fishingHookThreshold)
+                    {
+                        // caught
+                        OnCaughtFish();
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Called when received entity properties from server.
+        /// </summary>
+        /// <param name="EntityID"></param>
+        /// <param name="prop"></param>
+        public void OnEntityProperties(int EntityID, Dictionary<string, Double> prop)
+        {
+            if(EntityID == playerEntityID)
+            {
+                // adjust auto attack cooldown for maximum attack damage
+                if (prop.ContainsKey("generic.attackSpeed"))
+                {
+                    if (attackSpeed != prop["generic.attackSpeed"])
+                    {
+                        attackSpeed = prop["generic.attackSpeed"];
+                        attackCooldownSecond = 1 / attackSpeed * (serverTPS / 20.0); // server tps will affect the cooldown
+                        attackCooldown = Convert.ToInt16(Math.Truncate(attackCooldownSecond / 0.1) + 1);
+                    }
+                }
+            }
+        }
+        
+        /// <summary>
+        /// Called when server sent a Time Update packet.
+        /// </summary>
+        /// <param name="WorldAge"></param>
+        /// <param name="TimeOfDay"></param>
+        public void OnTimeUpdate(long WorldAge, long TimeOfDay)
+        {
+            if (!Settings.AutoAttackMobs) return;
+            // calculate server tps for adjusting attack cooldown
+            if (lastAge != 0)
+            {
+                DateTime currentTime = DateTime.Now;
+                Double tps = (WorldAge - lastAge) / (currentTime - lastTime).TotalSeconds;
+                lastAge = WorldAge;
+                lastTime = currentTime;
+                if (tps <= 20 || tps >= 0)
```

Comment:
Seems like this condition will always be true, did you mean `&&` instead of `||`?

Reply:
Actually the tps calculation is not very accurate (I don’t know why) and sometimes exceeds the range of 0-20

Reply:
Seems like you resolved it anyway in newer commits 😉